Set default MemorySwap on Docker side

hqhq opened the pull request:
As discussed in https://github.com/docker/libcontainer/issues/616,
we should move this behavior to Docker side.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

dqminh commented on daemon/daemon.go:
```diff
@@ -1168,6 +1168,13 @@ func checkKernel() error {
+		hostConfig.MemorySwap = hostConfig.Memory * 2
+	}
+}
+
```
i think we can merge this into `verifyContainerSettings` func

hqhq replied:
I'm not sure, they seems to be two different things, I'll merge if more comments agree this. Thanks.

rjnagal replied:
I'd say they are different things like @hqhq said. verifyContainerSettings() only sanitizes hostconfig when certain features are not available. This is more of policy settings.

phemmer commented:
I'm going to echo one of the sentiments made by @vmarmol over on docker/libcontainer#616

> I didn't ask for swap, yet libcontainer decided to default it to something even though there was no support.

Even if it only does this when libcontainer has support, I still didn't ask for it.
And the formula of `swap=mem*2` is EXTREMELY antiquated, and hasn't been relevant for decades.

dqminh commented:
> Even if it only does this when libcontainer has support, I still didn't ask for it.
And the formula of swap=mem*2 is EXTREMELY antiquated, and hasn't been relevant for decades.

I think I agree with this. It's probably better if memsw is only set when there's support and user ask for it.

hqhq commented:
I don't mind take this logic away, if users don't set it, leave it as unlimited seems more reasonable to me, just it'll break usage, maybe we can send a tip for a while then remove it in a smooth way.

rjnagal commented:
I would vote for just moving the logic from libcontainer to docker and stay consistent with the behavior users expect. We can do a more publicized deprecation later.

dqminh commented on daemon/create.go:
```diff
@@ -19,6 +19,7 @@ func (daemon *Daemon) ContainerCreate(name string, config *runconfig.Config, hos
 		return "", nil, fmt.Errorf("Config cannot be empty in order to create a container")
 	}
 
+	daemon.adaptContainerSettings(hostConfig)
```
i think this call should be set after `verifyContainerSettings` so when the host doesnt have `memsw` support, we dont set it to x2 mem and then set to -1 after with a warning.

hqhq replied:
Yes this setting *2 is redundant if we don't have `memsw` support, but this warning will show up either we put this function before or after. And I thought it'll be safer if we can verify configs after adapt them, maybe it'll avoid some mistakes in the future when `adaptContainerSettings` grows up. Set memsw *2 then set it to -1 doesn't hurt anyone, WDYT?

dqminh replied:
Ah yes, i think you are correct. The current code will show the warning even if user doesn't specify `MemorySwap` (i.e. MemorySwap == 0 ). Preserving this for now is ok for me.

calavera commented:
LGTM but it needs a rebase :pensive:

hqhq commented:
Sorry for the late, updated.

hqhq commented:
ping.

thaJeztah commented:
Got a panic on Janky and Experimental

hqhq committed:
Set default MemorySwap on Docker side

As discussed in docker/libcontainer#616,
we should move this behavior to Docker side.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@thaJeztah Fixed, all green now. :smile:

thaJeztah commented:
@hqhq thanks!

icecrime commented:
Code LGTM. Is it good for you this way @phemmer @rjnagal?

jfrazelle commented:
reping @phemmer @rjnagal dont want to merge this unless its ok

phemmer commented:
> I would vote for just moving the logic from libcontainer to docker and stay consistent with the behavior users expect. We can do a more publicized deprecation later.

I'm good with that.

jfrazelle commented:
cool LGTM

hqhq commented:
@jfrazelle Since @rjnagal approved this approach in the first place, @phemmer 's approve would be all we need, so I think we can go now, WDYT?

jfrazelle commented:
sorry, was late, forgot to push button ;) thanks!
